RVPS value extraction script Modification to handle few cases.

gauravkuredhat opened the pull request:
Description: The script needs to be modified to find 'nbd' device which is available. Also in case , HKD.crt has some issue then the script needs to produce appropriate error.


Please provide the following information:
-->
**- Description of the problem which is fixed/What is the use case**
1. There were a few cases where the HKD was incorrect, and script was giving blank values for parameters. This took long time to debug.
2. The script is using 'nbd3' as a default network block device which should be changed. In case 'nbd3' is not available, the script should use next available device.

More details are here -
https://issues.redhat.com/browse/KATA-3495

**- What I did**
Applied necessary changes to cater these 2.

**- How to verify it**
1. Run the script
2. Check extraction of RVPS from image which is there in local pc.
4. Check extraction of RVPS from iamge which is there in volume.
5. Try to give incorrect HKD.crt and check if it is giving proper error message.

**- Description for the changelog**
RVPS script enhancement.

gauravkuredhat commented:
Here is the result of the script after changes - 
```
1) Case with correct HKD.crt :- 
root@a3elp61 Rvps-Extraction]# ./GetRvps.sh
***Installing necessary packages for RVPS values extraction ***
Updating Subscription Management repositories.
Last metadata expiration check: 4:11:02 ago on Tue Dec  3 09:04:09 2024.
Package python3-3.9.19-8.el9_5.1.s390x is already installed.
Package python3-cryptography-36.0.1-4.el9.s390x is already installed.
Package kmod-28-10.el9.s390x is already installed.
Dependencies resolved.
Nothing to do.
Complete!
***Installation Finished ***
1) Generate the RVPS From Local Image from User pc  3) Quit
2) Generate RVPS from Volume
Please enter your choice: 1
Enter the Qcow2 image with Full path
/root/Rvps-Extraction/sandboxed-containers-operator/hack/Rvps-Extraction/PODVM-VOL-IMAGE/podvm_test.qcow2
nbd0 looks available. Starting Mounting the image with it..
mount: /mnt/myvm: special device /dev/nbd0p1 does not exist.
Error: Failed to mount the image. Retrying...
Mounting on second attempt passed
/dev/nbd0 disconnected
SE header found at offset 0x014000
SE header written to '/root/Rvps-Extraction/sandboxed-containers-operator/hack/Rvps-Extraction/output-files/hdr.bin' (2880 bytes)
se.tag:  8cb40fab7a2405ec228f67af9d350cce
se.image_phkh:  92d0aff6eb86719b6b1ea0cb98d2c99ff2ec693df3efff2158f54112f6961508
provenance = ewogICAgInNlLmF0dGVzdGF0aW9uX3Boa2giOiBbCiAgICAgICAgIjkyZDBhZmY2ZWI4NjcxOWI2YjFlYTBjYjk4ZDJjOTlmZjJlYzY5M2RmM2VmZmYyMTU4ZjU0MTEyZjY5NjE1MDgiCiAgICBdLAogICAgInNlLnRhZyI6IFsKICAgICAgICAiOGNiNDBmYWI3YTI0MDVlYzIyOGY2N2FmOWQzNTBjY2UiCiAgICBdLAogICAgInNlLmltYWdlX3Boa2giOiBbCiAgICAgICAgIjkyZDBhZmY2ZWI4NjcxOWI2YjFlYTBjYjk4ZDJjOTlmZjJlYzY5M2RmM2VmZmYyMTU4ZjU0MTEyZjY5NjE1MDgiCiAgICBdLAogICAgInNlLnVzZXJfZGF0YSI6IFsKICAgICAgICAiMDAiCiAgICBdLAogICAgInNlLnZlcnNpb24iOiBbCiAgICAgICAgIjI1NiIKICAgIF0KfQo=
-rw-r--r--. 1 root root 2880 Dec  3 13:15 /root/Rvps-Extraction/sandboxed-containers-operator/hack/Rvps-Extraction/output-files/hdr.bin
-rw-r--r--. 1 root root  561 Dec  3 13:15 /root/Rvps-Extraction/sandboxed-containers-operator/hack/Rvps-Extraction/output-files/se-message
-rw-r--r--. 1 root root  446 Dec  3 13:15 /root/Rvps-Extraction/sandboxed-containers-operator/hack/Rvps-Extraction/output-files/ibmse-policy.rego
```

2.Result with incorrect HKD.crt -
```
***Installation Finished ***
1) Generate the RVPS From Local Image from User pc
2) Generate RVPS from Volume
3) Quit
Please enter your choice: 1
Enter the Qcow2 image with Full path
/root/hpcc-se-initdata-working-v1.qcow2
nbd0 looks available. Starting Mounting the image with it..
mount: /mnt/myvm: special device /dev/nbd0p1 does not exist.
Error: Failed to mount the image. Retrying...
Mounting on second attempt passed
/dev/nbd0 disconnected
SE header found at offset 0x014000
SE header written to '/root/Rvps-Extraction/sandboxed-containers-operator/hack/Rvps-Extraction/output-files/hdr.bin' (2880 bytes)
Traceback (most recent call last):
  File "/root/Rvps-Extraction/sandboxed-containers-operator/hack/Rvps-Extraction/static-files/se_parse_hdr.py", line 162, in <module>
    parse_hdr(hdr_file, hkd_file)
  File "/root/Rvps-Extraction/sandboxed-containers-operator/hack/Rvps-Extraction/static-files/se_parse_hdr.py", line 85, in parse_hdr
    hkd_phkh = parse_img_phkh_from_hkd(hkd_file)
  File "/root/Rvps-Extraction/sandboxed-containers-operator/hack/Rvps-Extraction/static-files/se_parse_hdr.py", line 59, in parse_img_phkh_from_hkd
    public_key = parse_certificate(filename)
  File "/root/Rvps-Extraction/sandboxed-containers-operator/hack/Rvps-Extraction/static-files/se_parse_hdr.py", line 19, in parse_certificate
    cert = x509.load_pem_x509_certificate(cert_data, default_backend())
  File "/usr/lib64/python3.9/site-packages/cryptography/x509/base.py", line 531, in load_pem_x509_certificate
    return rust_x509.load_pem_x509_certificate(data)
ValueError: error parsing asn1 value: ParseError { kind: ShortData }
Traceback (most recent call last):
  File "/root/Rvps-Extraction/sandboxed-containers-operator/hack/Rvps-Extraction/static-files/se_parse_hdr.py", line 162, in <module>
    parse_hdr(hdr_file, hkd_file)
  File "/root/Rvps-Extraction/sandboxed-containers-operator/hack/Rvps-Extraction/static-files/se_parse_hdr.py", line 85, in parse_hdr
    hkd_phkh = parse_img_phkh_from_hkd(hkd_file)
  File "/root/Rvps-Extraction/sandboxed-containers-operator/hack/Rvps-Extraction/static-files/se_parse_hdr.py", line 59, in parse_img_phkh_from_hkd
    public_key = parse_certificate(filename)
  File "/root/Rvps-Extraction/sandboxed-containers-operator/hack/Rvps-Extraction/static-files/se_parse_hdr.py", line 19, in parse_certificate
    cert = x509.load_pem_x509_certificate(cert_data, default_backend())
  File "/usr/lib64/python3.9/site-packages/cryptography/x509/base.py", line 531, in load_pem_x509_certificate
    return rust_x509.load_pem_x509_certificate(data)
ValueError: error parsing asn1 value: ParseError { kind: ShortData }
se.tag:
se.image_phkh:
There seems to be some issue in HKD.crt. Please use the correct one and run it again.
```

snir911 commented:
Thanks! I'd squash all the commits together (or at least the 2 latter ones)
Also https://github.com/openshift/sandboxed-containers-operator/pull/479 is acked already and should be merged so it would be better to adapt this one to the new file locations

gauravkuredhat commented:
> Thanks! I'd squash all the commits together (or at least the 2 latter ones) Also #479 is acked already and should be merged so it would be better to adapt this one to the new file locations

==> Thanks @snir911 for the review. I have changed it.

gkurz commented:
Hi @gauravkuredhat ! You need to rebase this PR.

gkurz commented:
Hi @gauravkuredhat !

We don't  want merge commits in PRs. Instead of merging `devel` into your PR, please `git rebase`  in your local repo, fix the conflicts and then `git push -f` to the PR. Thanks !

gauravkuredhat commented:
> Hi @gauravkuredhat !
> 
> We don't want merge commits in PRs. Instead of merging `devel` into your PR, please `git rebase` in your local repo, fix the conflicts and then `git push -f` to the PR. Thanks !

==> sure @gkurz , I will check and update

gkurz commented:
Thanks @gauravkuredhat ! It looks much better but there are still at least two things to do:
- provide a log message in the commit (see 1c332562256c4d59dff59644ea282d17aac59263 for an example of log message). We want every commits in the repo to contain a description. This allows contributors to have all the important info at hand in their local repo without needing to go to github.
- provide a valid Signed-off-by. This is required by the Developer Certificate of Origin.

gauravkuredhat commented:
> Thanks @gauravkuredhat ! It looks much better but there are still at least two things to do:
> 
> * provide a log message in the commit (see [1c33256](https://github.com/openshift/sandboxed-containers-operator/commit/1c332562256c4d59dff59644ea282d17aac59263) for an example of log message). We want every commits in the repo to contain a description. This allows contributors to have all the important info at hand in their local repo without needing to go to github.
> * provide a valid Signed-off-by. This is required by the Developer Certificate of Origin.

==> Thanks @gkurz for the review. I have committed with above. Can you please check once.

gkurz commented on scripts/rvps-extraction/GetRvps.sh:
```diff
@@ -3,60 +3,107 @@
+
+	      sleep 2
+	      #manual creation of nbdxp1 using mknod
+	      mknod /dev/"$nbd_available"p1 b $devmajor $newdevminor
```
From the manual page of `qemu-nbd` :

```
Expose  the  guest-visible  contents of a qcow2 file via a block device
/dev/nbd0 (and possibly creating /dev/nbd0p1 and friends for partitions
found within), then disconnect the device when done.
[...]
qemu-nbd -c /dev/nbd0 -f qcow2 file.qcow2
```

Why are you trying to create the device node manually ? How do you expect this would help ?

gauravkuredhat replied:
Hi @gkurz,

We encountered a few cases where nbd0 is engaged, resulting in errors when attempting to use it. For example:

Refer to the case below:
[Documentation Link](https://docs.redhat.com/en/documentation/openshift_sandboxed_containers/1.8/html/user_guide/deploying-cc_ibm-cc#ibm-cc-obtaining-attest-fields-certs-keys_ibm-cc:~:text=%24%20lsblk-,Example%20output,-nbd0%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2043%3A0)

In this example, we can see that nbd0p1 is engaged by luks:

nbd0                                           43:0    0  100G  0 disk
├─nbd0p1                                       43:1    0  255M  0 part
├─nbd0p2                                       43:2    0    6G  0 part
│ └─luks-e23e15fa-9c2a-45a5-9275-aae9d8e709c3 253:2    0    6G  0 crypt
└─nbd0p3                                       43:3    0 12.4G  0 part
nbd1                                           43:32   0   20G  0 disk
├─nbd1p1                                       43:33   0  255M  0 part
├─nbd1p2                                       43:34   0    6G  0 part
│ └─luks-5a540f7c-c0cb-419b-95e0-487670d91525 253:3    0    6G  0 crypt
└─nbd1p3                                       43:35   0 86.9G  0 part
nbd2                                           43:64   0    0B  0 disk
nbd3                                           43:96   0    0B  0 disk
nbd4                                           43:128  0    0B  0 disk
nbd5                                           43:160  0    0B  0 disk
nbd6                                           43:192  0    0B  0 disk
nbd7                                           43:224  0    0B  0 disk
nbd8                                           43:256  0    0B  0 disk
nbd9                                           43:288  0    0B  0 disk
nbd10                                          43:320  0    0B  0 disk

With the new logic, the next available nbd device is nbd2, and it will be used.
If nbd0 becomes available, it will certainly be used.

Additionally, we faced some cases where nbd0p1 could not be used due to a partition error. In such scenarios, the script attempts to handle the issue by manually creating the partition. I’ve implemented the following logic:

1. Attempt to mount: mount /dev/"$nbd_available"p1 /mnt/myvm (This is the first attempt).
2. Retry: If step 1 fails, the script retries once.
3. Create manually: If the retrial also fails (a rare occurrence), the script creates the device using mknod /dev/"$nbd_available"p1 b $devmajor $newdevminor.
4. If still it fails, script will print a message to try LPAR reboot once.

In most cases, step 1 works. In a few cases, we encountered step 2, so I added a retry mechanism. Finally, in one isolated scenario on an LPAR, we faced step 3, so I included that logic as well.

Essentially, I have accounted for all these scenarios to ensure the script helps users retrieve the required values seamlessly.

Please let me know if this addresses your query. I will share additional details if needed.

gkurz replied:
> Hi @gkurz,
> 
> We encountered a few cases where nbd0 is engaged, resulting in errors when attempting to use it. For example:
> 
> [...]
> 
> With the new logic, the next available nbd device is nbd2, and it will be used. If nbd0 becomes available, it will certainly be used.
> 

As already stated in https://github.com/openshift/sandboxed-containers-operator/pull/486#pullrequestreview-2498995978, I'm fine with this part.

> Additionally, we faced some cases where nbd0p1 could not be used due to a partition error. In such scenarios, the script attempts to handle the issue by manually creating the partition. I’ve implemented the following logic:
> 
> 1. Attempt to mount: mount /dev/"$nbd_available"p1 /mnt/myvm (This is the first attempt).
> 2. Retry: If step 1 fails, the script retries once.
> 3. Create manually: If the retrial also fails (a rare occurrence), the script creates the device using mknod /dev/"$nbd_available"p1 b $devmajor $newdevminor.
> 

This doesn't make sense. The `mknod` command simply creates a block device special file, which is basically an interface between userspace and the nbd kernel driver. It will not create a partition in any way.

> In most cases, step 1 works. In a few cases, we encountered step 2, so I added a retry mechanism. Finally, in one isolated scenario on an LPAR, we faced step 3, so I included that logic as well.
> 

The nbd server runs in userspace. The on-lining of the nbd block device is asynchronous. It is certainly possible that it takes some time to complete inside an overloaded LPAR.

> Essentially, I have accounted for all these scenarios to ensure the script helps users retrieve the required values seamlessly.
> 

Step 3 is a no go. The only thing to do is to try again until the `/dev/nbdXp1`  eventually shows up or bail out after a certain amount of time.

> Please let me know if this addresses your query. I will share additional details if needed.

gauravkuredhat replied:
Thanks @gkurz . I have removed that part.

Result looks good to me. I am sharing below - 

Script Result from PODVM volume image :- 
```
1) Generate the RVPS From Local Image from User pc
2) Generate RVPS from Volume
3) Quit
Please enter your choice: 2
Enter the Libvirt Pool Name
fa-pp
Enter the Libvirt URI Name
qemu+ssh://root@192.168.122.1/system?no_verify=1
Enter the Libvirt Volume Name
fa-pp-vol
Downloading from PODVM Volume...

nbd0 looks available. Starting Mounting the image with it..
/dev/nbd0 disconnected
SE header found at offset 0x014000
SE header written to '/root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/hdr.bin' (640 bytes)
se.tag:  42f3fe61e8a7e859cab3bb033fd11c61
se.image_phkh:  92d0aff6eb86719b6b1ea0cb98d2c99ff2ec693df3efff2158f54112f6961508
provenance = ewogICAgInNlLmF0dGVzdGF0aW9uX3Boa2giOiBbCiAgICAgICAgIjkyZDBhZmY2ZWI4NjcxOWI2YjFlYTBjYjk4ZDJjOTlmZjJlYzY5M2RmM2VmZmYyMTU4ZjU0MTEyZjY5NjE1MDgiCiAgICBdLAogICAgInNlLnRhZyI6IFsKICAgICAgICAiNDJmM2ZlNjFlOGE3ZTg1OWNhYjNiYjAzM2ZkMTFjNjEiCiAgICBdLAogICAgInNlLmltYWdlX3Boa2giOiBbCiAgICAgICAgIjkyZDBhZmY2ZWI4NjcxOWI2YjFlYTBjYjk4ZDJjOTlmZjJlYzY5M2RmM2VmZmYyMTU4ZjU0MTEyZjY5NjE1MDgiCiAgICBdLAogICAgInNlLnVzZXJfZGF0YSI6IFsKICAgICAgICAiMDAiCiAgICBdLAogICAgInNlLnZlcnNpb24iOiBbCiAgICAgICAgIjI1NiIKICAgIF0KfQo=
-rw-r--r--. 1 root root 640 Dec 13 12:13 /root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/hdr.bin
-rw-r--r--. 1 root root 446 Dec 13 12:13 /root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/ibmse-policy.rego
-rw-r--r--. 1 root root 561 Dec 13 12:13 /root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/se-message
```



Script Result from PODVM local image :-

```
1) Generate the RVPS From Local Image from User pc
2) Generate RVPS from Volume
3) Quit
Please enter your choice: 1
Enter the Qcow2 image with Full path
/root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/PODVM-VOL-IMAGE/podvm_test.qcow2
nbd0 looks available. Starting Mounting the image with it..
mount: /mnt/myvm: special device /dev/nbd0p1 does not exist.
Error: Failed to mount the image. Retrying...
Mounting on second attempt passed
/dev/nbd0 disconnected
SE header found at offset 0x014000
SE header written to '/root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/hdr.bin' (640 bytes)
se.tag:  42f3fe61e8a7e859cab3bb033fd11c61
se.image_phkh:  92d0aff6eb86719b6b1ea0cb98d2c99ff2ec693df3efff2158f54112f6961508
provenance = ewogICAgInNlLmF0dGVzdGF0aW9uX3Boa2giOiBbCiAgICAgICAgIjkyZDBhZmY2ZWI4NjcxOWI2YjFlYTBjYjk4ZDJjOTlmZjJlYzY5M2RmM2VmZmYyMTU4ZjU0MTEyZjY5NjE1MDgiCiAgICBdLAogICAgInNlLnRhZyI6IFsKICAgICAgICAiNDJmM2ZlNjFlOGE3ZTg1OWNhYjNiYjAzM2ZkMTFjNjEiCiAgICBdLAogICAgInNlLmltYWdlX3Boa2giOiBbCiAgICAgICAgIjkyZDBhZmY2ZWI4NjcxOWI2YjFlYTBjYjk4ZDJjOTlmZjJlYzY5M2RmM2VmZmYyMTU4ZjU0MTEyZjY5NjE1MDgiCiAgICBdLAogICAgInNlLnVzZXJfZGF0YSI6IFsKICAgICAgICAiMDAiCiAgICBdLAogICAgInNlLnZlcnNpb24iOiBbCiAgICAgICAgIjI1NiIKICAgIF0KfQo=
-rw-r--r--. 1 root root 640 Dec 13 12:14 /root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/hdr.bin
-rw-r--r--. 1 root root 446 Dec 13 12:14 /root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/ibmse-policy.rego
-rw-r--r--. 1 root root 561 Dec 13 12:14 /root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/se-message
```

gkurz replied:
Thanks @gauravkuredhat ! This is better. Now I'd like you to :
- squash these two patches into a single commit in your local repo
- get rid of `1. The script should provide alternates in case there will be some issue on nbd subdevices (e.g :- nbd0p1 is not available)`  in the commit log
- force push to the PR

For next time, please avoid adding partial updates to your PR.  You should process requests from the review in your local repo and force push to GH.

gauravkuredhat replied:
Hi @gkurz , done. Please can you check.

gkurz replied:
@gauravkuredhat please add commit log and Signed-off-by.

gauravkuredhat replied:
Done @gkurz

gkurz reviewed (commented):
Hi @gauravkuredhat !

> 1. The script should provide alternates in case there will be some issue
   on nbd subdevices (e.g :- nbd0p1 is not available)

I must admit that I don't quite understand the logic. The `nbd` subdevices are supposed to be created according to the content of the image that `qemu-nbd` is serving... `nbd0p1` not showing up is likely because the image doesn't contain such a partition, in which case the only thing to do is to complain that the image is broken and exit.

> 2. In case the nbdx is engaged, the script should find next available
   nbd and try to mount with it

This part is fine.

gkurz commented:
Converted to draft until https://github.com/openshift/sandboxed-containers-operator/pull/486#discussion_r1881862446 is sorted out.

gkurz reviewed (approved):
/lgtm

Thanks @gauravkuredhat !

snir911 commented on scripts/rvps-extraction/GetRvps.sh:
```diff
@@ -3,60 +3,75 @@
+
+    #Check the available nbd and use it
+    nbd_available=`lsblk | grep nbd | grep 0B | awk '{print $1}' | head -1`
+    echo $nbd_available "looks available. Starting Mounting the image with it.."
```
maybe worth checking the string exist (and exit if not?) as it may print the msg also when it's nor available

gkurz replied:
Good catch @snir911 ! It will indeed print a `  looks available. Starting Mounting the image with it..` message and `qemu-nbd -c /dev/ $img_path` will then fail right away with `qemu-nbd: Failed to open /dev: Is a directory`. Testing the string isn't empty will allow to come up with a more meaningful error message.

Please process this comment @gauravkuredhat.

gauravkuredhat replied:
Changed @snir911 @gkurz . thanks

snir911 reviewed (approved):
LGTM overall, thanks, pls see minor comment

gauravkuredhat committed:
Changes for Enhancement in RVPS value extraction script. This includes:-
1. Check for available nbd device and mount it rather than using
   hardcoded values. In case nbd does not exist, script should give
proper message.

2. Add cases for handling proper log messages when incorrect HKD.crt is
   being used.
Changes for nbd availability negative test case

Signed-off-by: Gaurav Kumar <gauravku@redhat.com>
